apps/prairielearn/src/ee/lib/ai-question-generation-benchmark.ts

```diff
@@ -0,0 +1,477 @@
+    const totalScore = scores.reduce((sum, score) => sum + score, 0);
+    const avgScore = totalScore / scores.length;
+    const totalPoints = scores.length * 5;
+
+    job.info('==================');
+    job.info('Benchmark complete');
+    job.info(`Min score: ${minScore}`);
+    job.info(`Max score: ${maxScore}`);
+    job.info(`Average score: ${avgScore}`);
+    job.info(`Overall score: ${totalScore}/${totalPoints} (${(totalScore / totalPoints) * 100}%)`);
+  });
+
+  return serverJob.jobSequenceId;
+}
+
+async function evaluateGeneratedQuestion({
+  client,
+  authnUserId,
+  originalSystemPrompt,
+  userPrompt,
+  html,
+  python,
+}: {
+  client: OpenAI;
+  authnUserId: string;
+  originalSystemPrompt: string | null;
+  userPrompt: string;
+  html: string;
+  python: string;
+}): Promise<QuestionGenerationEvaluation | null> {
+  const systemPrompt = [
+    'Another LLM has generated a PrairieLearn question from the following prompt:',
+    '',
+    '<prompt>',
+    userPrompt
+      .split('\n')
+      .map((line) => line.trim())
+      .join('\n'),
+    '</prompt>',
+    '',
+    'Please evaluate it for correctness and clarity.',
+    'If anything is incorrect or could be improved, please provide feedback.',
+    'If the question looks fine, please indicate that as well.',
+    '',
+    'Do not suggest ways the original prompt could have been improved.',
+    'You should evaluate the question based on the prompt as it is written.',
+    '',
+    'Remember that for pedagogical reasons, instructors may include extra information or distractors.',
+    'They may also choose to make students carefully consider unit conversions without explicitly prompting for them.',
+    'These are not considered errors unless they are factually incorrect or egregiously misleading; do not factor them into your evaluation.',
+    '',
+    // This should include relevant documentation for the elements used in the
+    // generated question, as well as oru baseline context about how PrairieLearn
+    // works and some example questions. This is a lot of tokens, but it's
+    // important for accurate evaluation since it'll tell the LLM about which
+    // elements/attributes are available and how elements behave.
+    originalSystemPrompt
+      ? [
+          'The following context was provided to the LLM when it was asked to generate the question:',
+          '<context>',
+          originalSystemPrompt,
+          '</context>',
+          '',
+        ]
+      : [],
+    'The user will now provide the question HTML and Python files that the LLM generated for this prompt.',
+  ]
+    .flat()
+    .join('\n');
+
+  const generatedQuestion: string[] = ['```html', html.trim(), '```'];
+  if (python) {
+    generatedQuestion.push('', '```python', python.trim(), '```');
+  } else {
+    generatedQuestion.push('', 'No Python file was generated.');
+  }
+
+  const completion = await client.beta.chat.completions.parse({
+    model: 'gpt-4o',
```

Comment:
Should we generally fix the model versions to improve our repeatability? Possibly in a config variable?
https://github.com/PrairieLearn/PrairieLearn/blob/cd42ba3408430914fba00d6b3a23ba574480bf98/apps/prairielearn/src/ee/lib/ai-grading.ts#L355

Reply:
idk if a config variable is needed, I doubt we'd want to chance this on the fly (outside of local dev, where we can just edit it). I'd be happy to pin to a particular model though.

Reply:
I pinned to `gpt-4o-2024-11-20` in fd29c99.

Reply:
Perfect!